[Profiler] Emit skipped regions for inactive `#if` branches

hamishknight opened the pull request:
For any `#if` blocks in the function we're emitting, emit skipped ranges for the inactive clauses, including the syntax for the `#if` itself, since that should not be considered executable code.

rdar://116860865

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test

ahoppen commented on include/swift/Basic/SourceManager.h:
```diff
@@ -92,6 +92,37 @@ class GeneratedSourceInfo {
+  /// The range of the entire \c #if block, including \c #else and \c #endif.
+  CharSourceRange WholeRange;
+
+  /// The range of the active selected body, if there is one.
```
Does the range include `#if` (or equivalent as well) or just the actual statements/expressions/declarations? Also, I assume that an invalid active range represents that there is no active branch, right? Might be worth pointing that out in the doc comment.

These are the things I wonder about since I’ve been working with syntax a lot recently 😉

hamishknight replied:
Yeah it's the body, not the syntax. I'll add a comment :)

ahoppen commented on lib/Basic/SourceLoc.cpp:
```diff
@@ -408,6 +408,66 @@ SourceManager::getGeneratedSourceInfo(unsigned bufferID) const {
   return known->second;
 }
 
+ArrayRef<CharSourceRange> IfConfigRangeInfo::getRangesWithoutActiveBody(
+    SourceManager &SM, SmallVectorImpl<CharSourceRange> &scratch) const {
```
Wouldn’t it be easier to just return `SmallVector<CharSourceRange, 2>` here? Avoids the scratch and the function has more knowledge that the vector will contain at most 2 elements.

hamishknight replied:
Sure, I don't really feel too strongly about it, although I should note that returning a `SmallVector<CharSourceRange, 2>` doesn't guarantee at most 2 elements :) I don't think it's a guarantee this function needs to make though, callers ought to be able to handle more than 2 elements.

ahoppen replied:
> The `N` here is just the number of in-place elements, I think @ahoppen's point is that in this case, `getRangesWithoutActiveBody` knows more about the likely number of elements then eg. a caller would.

Yes, exactly.

ahoppen commented on lib/Basic/SourceLoc.cpp:
```diff
@@ -408,6 +408,66 @@ SourceManager::getGeneratedSourceInfo(unsigned bufferID) const {
+  assert(scratch.empty());
+  if (ActiveBodyRange.isValid()) {
+    // Split the whole range by the active range.
+    scratch.emplace_back(SM, WholeRange.getStart(), ActiveBodyRange.getStart());
+    scratch.emplace_back(SM, ActiveBodyRange.getEnd(), WholeRange.getEnd());
```
Not sure if this could cause issue but should we only add the ranges if they are not empty?

hamishknight replied:
`WholeRange` should definitely never be empty, I believe `ActiveBodyRange` _could_ be empty, but only in invalid code (we error on e.g `#if true#endif`), but it shouldn't make a difference here, i.e we want the entire range in that case anyway. I'll at least add an assert for `WholeRange` though.

ahoppen replied:
But given https://github.com/apple/swift/pull/69782#discussion_r1392645561, looks like this will never come up in practice anyway.

hamishknight replied:
Ah right, I don't think that can happen, since the active range never includes the syntax of the `#if` and the whole range does. I guess we could add a check to be defensive though

hamishknight replied:
Oh heh only just saw your latest comment :)

ahoppen reviewed (commented):
I only reviewed how we store the active `#if` ranges.

---

Does this mean we can get rid of `IfConfigDecl` if we store active `#if` branches in the source manager?

---

I’m not sure how I feel about storing the information in `SourceManager`. It kind of infiltrates parsing concepts into the `SourceManager` which should be at a lower level. Wouldn’t `ASTContext` be a better fit to store this? It looks like the only other parsing-related information that we store in `SourceManager` are the `RegexLiteralStartLocs` and they have the following comment

```cpp
  /// The starting source locations of regex literals written in source. This
  /// is an unfortunate hack needed to allow for correct re-lexing.
```

bnbarham commented:
> I’m not sure how I feel about storing the information in SourceManager. It kind of infiltrates parsing concepts into the SourceManager which should be at a lower level. Wouldn’t ASTContext be a better fit to store this?

Or eg. `SourceFile`. We have a `ModuleSourceFileLocationMap` in `ModuleDecl` to find a `SourceFile` for a given `SourceLocation` (`getSourceFileContainingLocation`). I believe that would allow removing all the sorting handling, since `ModuleSourceFileLocationMap` already handles it for across files and within a file the regions should already be sorted (?).

hamishknight commented:
> Does this mean we can get rid of `IfConfigDecl` if we store active `#if` branches in the source manager?

At some point we probably could, we need to migrate clients like MiscDiagnostics off it though, as it's currently walking the inactive bodies for VarDecl usage checking.

> I’m not sure how I feel about storing the information in `SourceManager`. It kind of infiltrates parsing concepts into the `SourceManager` which should be at a lower level. Wouldn’t `ASTContext` be a better fit to store this?

Yeah, I think I agree with Ben that the SourceFile is probably a better place for this. I didn't put too much consideration into this initially since this is hopefully something we can eventually replace with a syntax tree query.

hamishknight commented:
@swift-ci please test

ahoppen commented on lib/SIL/Parser/ParseSIL.cpp:
```diff
@@ -8621,13 +8621,22 @@ bool SILParserState::parseSILCoverageMap(Parser &P) {
-        break;
+      using MappedRegion = SILCoverageMap::MappedRegion;
+
+      if (P.Tok.is(tok::identifier) && P.Tok.getText().equals("skipped")) {
```
I think checking for contextual keyword would be better here. Not sure if SIL has escaped identifiers but this will check that we didn’t write backticks around `skipped`.

```suggestion
      if (P.Tok.isContextualKeyword("skipped")) {
```

hamishknight replied:
Ah nice

hamishknight commented:
@swift-ci please test

hamishknight committed:
[Profiler] Emit skipped regions for inactive `#if` branches

For any `#if` blocks in the function we're emitting,
emit skipped ranges for the inactive clauses, including
the syntax for the `#if` itself, since that should
not be considered executable code.

rdar://116860865

hamishknight committed:
[Parse] Store `#if` ranges on the SourceFile

Record any `#if`'s we encounter, and make a note
of their active region.

hamishknight committed:
[SIL] Introduce skipped coverage ranges

Introduce a skipped region kind, printed using
a `skipped` keyword.

hamishknight commented:
@swift-ci please test
